### PR TITLE
fix: Behavior form should prompt anytime you navigate away and there are changes pending

### PR DESCRIPTION
**Pull request by msnyder-msft**
### Summary of changes 🔍 

Moving all of the state management for the behavior form into a context. This allows us to track every update that is made to the behavior and know whether it has any changes or not. With this information, we can then show a confirmation dialog anytime the user tries to navigate away from the form. These flows include:
- cancel form
- mode switch toggle (viewer/builder)
- breadcrumb navigation

We do NOT show the prompt in any of those cases if no changes have been made OR if the changes were reverted back to matching the original state (ex: add an element and then remove it, then try to navigate, would not get a confirmation)

image

[Product Backlog Item 14504805](https://msazure.visualstudio.com/One/_workitems/edit/14504805): Dirty state prompt should be handled more holistically

Fixes issue #472 

**TODO**: Add more unit tests for the context

### Testing 🧪
Try everything related to the behavior form and make sure that any change in the form would result in a dialog before navigating away from the form.

Added unit tests for the context
image

### Checklist ✔️
- [x] Linked associated issue (if present)
- [x] Added relevant labels
- [x] Requested developer reviews
- [x] Request UI review from design / PM
- [x] Verify all code checks are passing

### CODE REVIEW

**Comment by ccrowley96** [comment]
The behaviors form alerts tab seems to be keeping edits to the draft even after clicking cancel & discard

**Comment by ccrowley96** [comment]
The live update behavior preview modal seems to be gone.  Not sure why, but def need that back haha

File: src/Components/ADT3DSceneBuilder/Internal/BuilderLeftPanel/BuilderLeftPanel.tsx [hunk 1]
```diff
@@ -445,9 +461,67 @@ const BuilderLeftPanel: React.FC<IBuilderLeftPanelProps> = ({ styles }) => {
...
+                    if (isDirty) {
+                        logDebugConsole(
+                            'debug',
+                            `[END] pre-breadcrumb navigation of action ${action} on ElementForm. {isDirty}`,
+                            isDirty
+                        );
+                        setUnsavedBehaviorChangesDialogOpen(true);
+                        setUnsavedChangesDialogDiscardAction(() => {
+                            navigate();
+                            dispatch({
```
Comment (msnyder-msft) [inline src/Components/ADT3DSceneBuilder/Internal/BuilderLeftPanel/BuilderLeftPanel.tsx:499]: remove

File: src/Components/SceneBreadcrumb/SceneBreadcrumbFactory.tsx [hunk 2]
```diff
@@ -63,52 +81,68 @@ const SceneBreadcrumbFactory: React.FC<ISceneBreadcrumbFactoryProps> = ({
...
-                            ADT3DSceneBuilderMode.EditBehavior)) && {
+                ...(isClickable && {
                     onClick: () => {
-                        cancelWidgetForm(widgetFormInfo, setWidgetFormInfo);
-                        setBehaviorTwinAliasFormInfo(null);
+                        const navigate = () => {
+                            cancelWidgetForm(widgetFormInfo, setWidgetFormInfo);
+                            setBehaviorTwinAliasFormInfo(null);
+                        };
+                        onNavigate('goToForm', navigate);
```
Comment (msnyder-msft) [inline src/Components/SceneBreadcrumb/SceneBreadcrumbFactory.tsx:112]: use new name

**Comment by ccrowley96** [comment]
When in a dirty state on the behaviors form, if you click on the scene name in the breadcrumb, the 'are you sure' dialog opens, but on clicking 'discard' nothing happens - you just stay on the widget form.  `widgetFormInfo` likely needs to be cleared out somewhere.

image

File: src/Components/ADT3DSceneBuilder/Internal/BuilderLeftPanel/BuilderLeftPanel.tsx (hunk 1 above)
Comment (msnyder-msft) [reply src/Components/ADT3DSceneBuilder/Internal/BuilderLeftPanel/BuilderLeftPanel.tsx:499]: done.

File: jest.setup.ts [hunk 3]
```diff
@@ -12,4 +14,11 @@ jest.mock('react-i18next', () => ({
         };
     }
 }));
-process.env.isTestMode = 'true'; // set this so we don't log in i18n
+
+// mock out utilities
+const MOCK_GUID = '00000000-1111-2222-3333-444444444444';
+jest.mock('./src/Models/Services/Utils', () => {
+    const actual = jest.requireActual('./src/Models/Services/Utils');
+    actual.createGuid = () => MOCK_GUID;
```
[the remark this answers is not in the document]
Comment (msnyder-msft) [reply jest.setup.ts]: removed.

File: src/Components/ADT3DSceneBuilder/Internal/Behaviors/Internal/AlertsTab.tsx [hunk 4]
```diff
@@ -73,85 +87,116 @@ const AlertsTab: React.FC = () => {
...
-                    );
-                    alertVisual.valueRanges[0].visual.color = newValue.item;
-                })
-            ),
-        [setBehaviorToEdit]
+            setValueRangeProperty('color', newValue.item),
+        [setValueRangeProperty]
     );
 
     const onIconChange = useCallback(
```
[the remark this answers is not in the document]
Comment (pasanchMSFT) [reply src/Components/ADT3DSceneBuilder/Internal/Behaviors/Internal/AlertsTab.tsx:172]: just send in the property they change as an argument, and they can be the same function.

**Review by ccrowley96 — APPROVED** [review]
Code looks great!  Super solid refactor : )

Left a couple of UX bug related comments.

**Comment by BrennaCourtney** [comment]
Editing an existing widget then clicking 'Cancel' doesn't prompt user with Discard message (if that behavior is covered under this PR)
- Edit an existing Widget (changing the value ranges or display name, etc.)
- Hit 'Cancel'
- Just exits you out of Widgets tab
- Click back in, changes gone

File: src/Components/ADT3DSceneBuilder/Internal/Behaviors/Internal/AlertsTab.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (msnyder-msft) [reply src/Components/ADT3DSceneBuilder/Internal/Behaviors/Internal/AlertsTab.tsx:172]: I think it's alright. we could remove the callbacks entirely and just pass it directly inline.

**Comment by msnyder-msft** [comment]
> The live update behavior preview modal seems to be gone. Not sure why, but def need that back haha

Should be fixed now.

**Comment by msnyder-msft** [comment]
> Editing an existing widget then clicking 'Cancel' doesn't prompt user with Discard message (if that behavior is covered under this PR)
> 
> * Edit an existing Widget (changing the value ranges or display name, etc.)
> * Hit 'Cancel'
> * Just exits you out of Widgets tab
> * Click back in, changes gone

This is expected for now. We are focused on Behaviors atm but it'd be great to get the widgets inline eventually as well.

**Comment by msnyder-msft** [comment]
> When in a dirty state on the behaviors form, if you click on the scene name in the breadcrumb, the 'are you sure' dialog opens, but on clicking 'discard' nothing happens - you just stay on the widget form. `widgetFormInfo` likely needs to be cleared out somewhere.
> 
> image
> 
> Update -- the same seems to be true anywhere from the behavior edit form. Clicking the scene name in the breadcrumb opens the dialog, but confirming the discard doesn't do anything

Good catch. I think I've fixed this now. I had a slight regression on a recent commit.

**Review by pasanchMSFT — APPROVED** [review]
Great changes, one quick note to take a look at in the future:
- If one or multiple items in elements list or layer list are unselected and then reselected, not changing the Behavior, prompt is opened. Edge case and doesn't really affect the user much, so still approving.

**Pull request merged**